File: services/dnk_canvas_api/supervisor/state_machine.py
# --- DNK-MRH-HEADER ---
# mrh_id: "supervisor/state_machine.py"
# purpose: "Implement extended state machine and valid state transitions for design runs."
# canonical_source: true
# status: "Active"
# version: "1.1.0"
# updated_at: "2026-08-11"
# author: "DNK-e.com Maksym"
# license: "DNK-INTERNAL"
# --- END DNK-MRH-HEADER ---

from typing import Set, Dict

class InvalidStateTransitionException(Exception):
    def __init__(self, from_state: str, to_state: str):
        self.from_state = from_state
        self.to_state = to_state
        super().__init__(f"409 INVALID_STATE_TRANSITION: Cannot transition from '{from_state}' to '{to_state}'")

class SupervisorStateMachine:
    # Set of valid states including new intermediate LLM workflow steps
    STATES = {
        "queued", "planning", "awaiting_context", "running", 
        "model_requested", "model_completed", "validating", "tool_pending",
        "waiting_approval", "materializing", "completed", "failed", 
        "cancelled", "expired"
    }

    # Dict of allowed transitions
    ALLOWED_TRANSITIONS: Dict[str, Set[str]] = {
        "queued": {"planning", "failed", "cancelled"},
        "planning": {"awaiting_context", "running", "failed", "cancelled"},
        "awaiting_context": {"running", "failed", "cancelled"},
        "running": {"model_requested", "waiting_approval", "materializing", "failed", "cancelled"},
        "model_requested": {"model_completed", "failed", "cancelled"},
        "model_completed": {"validating", "failed", "cancelled"},
        "validating": {"tool_pending", "materializing", "failed", "cancelled"},
        "tool_pending": {"materializing", "waiting_approval", "failed", "cancelled"},
        "waiting_approval": {"running", "tool_pending", "cancelled", "failed", "completed"},
        "materializing": {"completed", "failed"},
        "failed": {"queued"},
        "cancelled": set(),
        "completed": set(),
        "expired": set()
    }

    @classmethod
    def validate_transition(cls, from_state: str, to_state: str):
        if from_state not in cls.STATES:
            raise InvalidStateTransitionException(from_state, to_state)
        if to_state not in cls.STATES:
            raise InvalidStateTransitionException(from_state, to_state)
            
        allowed = cls.ALLOWED_TRANSITIONS.get(from_state, set())
        if to_state not in allowed and from_state != to_state:
            raise InvalidStateTransitionException(from_state, to_state)

File: packages/video-audit-core/src/pipeline/domain/jobs/state-machine.ts
/*
# --- DNK-MRH-HEADER ---
# mrh_id: "packages/video-audit-core/src/pipeline/domain/jobs/state-machine.ts"
# purpose: "Deterministic Finite State Machine & Transition Rules for Video Audit Jobs."
# canonical_source: true
# status: "Active"
# version: "1.0.0"
# updated_at: "2026-09-02"
# author: "DNK-e.com Maksym"
# --- END DNK-MRH-HEADER ---
*/

import { AuditJobStatus, VideoAuditJob } from './job.js';

export class InvalidJobTransitionError extends Error {
  public readonly fromStatus: AuditJobStatus;
  public readonly toStatus: AuditJobStatus;
  public readonly jobId: string;

  constructor(jobId: string, fromStatus: AuditJobStatus, toStatus: AuditJobStatus, reason?: string) {
    const msg = `Invalid state transition for job '${jobId}': cannot transition from '${fromStatus}' to '${toStatus}'.${
      reason ? ` Reason: ${reason}` : ''
    }`;
    super(msg);
    this.name = 'InvalidJobTransitionError';
    this.jobId = jobId;
    this.fromStatus = fromStatus;
    this.toStatus = toStatus;
    Object.setPrototypeOf(this, InvalidJobTransitionError.prototype);
  }
}

export interface JobTransitionRecord {
  jobId: string;
  fromStatus: AuditJobStatus;
  toStatus: AuditJobStatus;
  actorId: string;
  timestamp: string;
  reason?: string;
  attempt: number;
}

/**
 * Strict state transition matrix:
 *
 * QUEUED         -> LEASED
 * LEASED         -> RUNNING
 * LEASED         -> QUEUED (lease expired / reclaimed)
 * RUNNING        -> SUCCEEDED
 * RUNNING        -> RETRYABLE_ERROR
 * RUNNING        -> PERMANENT_ERROR
 * RUNNING        -> CANCELLED
 * RETRYABLE_ERROR-> QUEUED (scheduled retry)
 *
 * Terminal states (no outgoing transitions allowed):
 * SUCCEEDED
 * PERMANENT_ERROR
 * CANCELLED
 */
export const ALLOWED_JOB_TRANSITIONS: Readonly<Record<AuditJobStatus, readonly AuditJobStatus[]>> = {
  queued: ['leased'],
  leased: ['running', 'queued'],
  running: ['succeeded', 'retryable_error', 'permanent_error', 'cancelled'],
  retryable_error: ['queued', 'permanent_error'],
  succeeded: [],
  permanent_error: [],
  cancelled: []
};

export function canTransition(from: AuditJobStatus, to: AuditJobStatus): boolean {
  const allowed = ALLOWED_JOB_TRANSITIONS[from];
  return Boolean(allowed && allowed.includes(to));
}

export function validateTransition(
  jobId: string,
  from: AuditJobStatus,
  to: AuditJobStatus,
  reason?: string
): void {
  if (!canTransition(from, to)) {
    throw new InvalidJobTransitionError(jobId, from, to, reason);
  }
}

export const validateJobTransition = validateTransition;

export function applyJobTransition(
  job: VideoAuditJob,
  toStatus: AuditJobStatus,
  actorId: string,
  timestamp: string,
  reason?: string
): { updatedJob: VideoAuditJob; transitionRecord: JobTransitionRecord } {
  validateTransition(job.id, job.status, toStatus, reason);

  const transitionRecord: JobTransitionRecord = {
    jobId: job.id,
    fromStatus: job.status,
    toStatus,
    actorId,
    timestamp,
    reason,
    attempt: job.attempt
  };

  const updatedJob: VideoAuditJob = {
    ...job,
    status: toStatus
  };

  if (toStatus === 'leased') {
    updatedJob.leasedBy = actorId;
  } else if (toStatus === 'running') {
    if (!updatedJob.startedAt) {
      updatedJob.startedAt = timestamp;
    }
  } else if (toStatus === 'succeeded' || toStatus === 'permanent_error' || toStatus === 'cancelled') {
    updatedJob.finishedAt = timestamp;
    updatedJob.leasedUntil = undefined;
    updatedJob.leasedBy = undefined;
  } else if (toStatus === 'queued') {
    updatedJob.leasedUntil = undefined;
    updatedJob.leasedBy = undefined;
  }

  return { updatedJob, transitionRecord };
}
